plutus-chain-index/src/Plutus/ChainIndex/SyncStats.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Plutus.ChainIndex.SyncStats where
+
+import Cardano.BM.Tracing (ToObject)
+import Control.Concurrent (threadDelay)
+import Control.Concurrent.STM (TChan, atomically, dupTChan, tryReadTChan)
+import Control.Lens (preview)
+import Control.Monad (forever)
+import Control.Monad.Freer (Eff, LastMember, Member)
+import Control.Monad.Freer.Extras (LogMsg, logInfo, logWarn)
+import Control.Monad.IO.Class (liftIO)
+import Data.Aeson (FromJSON, ToJSON)
+import GHC.Generics (Generic)
+import Ledger (Slot (Slot))
+import Plutus.ChainIndex (tipAsPoint)
+import Plutus.ChainIndex qualified as CI
+import Plutus.ChainIndex.Lib (ChainSyncEvent (Resume, RollBackward, RollForward))
+import Prettyprinter (Pretty (pretty), comma, (<+>))
+import Text.Printf (printf)
+
+data SyncStats = SyncStats
+    { syncStatsAppliedBlocks    :: Integer -- ^ Number of applied blocks
+    , syncStatsAppliedRollbacks :: Integer -- ^ Number of rollbacks
+    , syncStatsChainSyncPoint   :: CI.Point -- ^ Chain index syncing tip
+    , syncStatsNodePoint        :: CI.Point -- ^ Current tip of the node
+    }
+    deriving stock (Eq, Show, Generic)
+    deriving anyclass (FromJSON, ToJSON, ToObject)
+
+instance Semigroup SyncStats where
+    SyncStats n1 m1 ct1 nt1 <> SyncStats n2 m2 ct2 nt2 =
+        SyncStats (n1 + n2) (m1 + m2) (ct1 <> ct2) (nt1 <> nt2)
+
+instance Monoid SyncStats where
+    mempty = SyncStats 0 0 mempty mempty
+
+data SyncLog =
+    NoNewBlocksSyncLog
+  | NewBlocksSyncLog Integer Integer CI.Point Slot
+    deriving stock (Eq, Show, Generic)
+    deriving anyclass (FromJSON, ToJSON, ToObject)
+
+instance Pretty SyncLog where
+  pretty = \case
+    NoNewBlocksSyncLog -> "No synced blocks in the last 30s."
+    NewBlocksSyncLog numRollForward numRollBackwards chainSyncPoint (Slot nodeSlot) ->
+        let (Slot chainSyncSlot) = maybe 0 fst $ preview CI._Point chainSyncPoint
+            pct = ((100 :: Double) * fromIntegral chainSyncSlot) / fromIntegral nodeSlot
+         in    "Syncing ("
+            <> pretty (printf "%.2f" pct :: String)
+            <> "%)"
+            <> ". Applied"
+            <+> pretty numRollForward
+            <+> "blocks"
+            <> comma
+            <+> pretty numRollBackwards
+            <+> "rollbacks in the last 30s. Current tip is"
+            <+> pretty chainSyncPoint
+
+-- | Read 'ChainSyncEvent's for the 'TChan' every 30 seconds and log syncing summary.
+logProgress :: forall effs.
+    ( Member (LogMsg SyncLog) effs
+    , LastMember IO effs
+    )
+    => TChan ChainSyncEvent
+    -> Eff effs ()
+logProgress broadcastChan = do
+    chan <- liftIO $ atomically $ dupTChan broadcastChan
+    forever $ do
+        liftIO $ threadDelay 30000000 -- 30s
+        (SyncStats a b chainSyncPoint nodePoint) <- liftIO $ generateSummary chan
+        case fmap fst $ preview CI._Point nodePoint of
+            Nothing -> do
+                logWarn NoNewBlocksSyncLog
```

Comment:
I guess you'd only want to see this while syncing, so only while `chainSyncPoint` has not reached `nodePoint`.

Reply:
Mmm.. good point. However, I think we would still want to see how many blocks were added even when `chainSyncPoint` reached `nodePoint`. When that happens, we could log every 300s instead of 30s? That's what `cardano-wallet` does.

Reply:
Sorry, I meant specifically the "no new blocks" message.